applications/nrf_desktop/src/modules/led_stream.c

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <zephyr.h>
+
+#define MODULE led_stream
+#include "module_state_event.h"
+
+#include <logging/log.h>
+
+#include "config_event.h"
+#include "led_event.h"
+#include "led_ready_event.h"
+
+#define STREAM_LED_ID 0
+#define QUEUE_SIZE 5
+
+LOG_MODULE_REGISTER(MODULE, CONFIG_DESKTOP_LED_STREAM_LOG_LEVEL);
+
+static struct led_effect led_stream_effect = {
+        .step_count = 1,
+};
+static struct led_effect_step steps_queue[QUEUE_SIZE];
+static s8_t last_added_position = -1;
+static s8_t send_position;
+static u8_t free_places = QUEUE_SIZE;
+static bool first_sent;
+static bool leds_initialized;
+
+static void queue_data(const struct event_dyndata *dyndata, u8_t position)
+{
+    LOG_INF("Insert_data_to_effect");
+    BUILD_ASSERT(CONFIG_DESKTOP_LED_COUNT == 3);
+    size_t pos = 0;
+    memcpy(steps_queue[position].color.c, &dyndata->data[pos], 3);
+    pos += 3;
+    steps_queue[position].substep_count=dyndata->data[pos];
+    pos += sizeof(steps_queue[position].substep_count);
+    steps_queue[position].substep_time=dyndata->data[pos];
+}
+
+static void send_effect(struct led_effect *effect)
+{
+    LOG_INF("Send_effect to leds");
+    struct led_event *led_event = new_led_event();
+    led_event->stream = true;
+    led_event->led_id = STREAM_LED_ID;
+    led_event->led_effect=effect;
+
+    __ASSERT_NO_MSG(led_event->led_effect->steps);
+
+    EVENT_SUBMIT(led_event);
+}
+
+static void store_data(const struct event_dyndata *data)
+{
+    if (free_places > 0) {
+        if (last_added_position == QUEUE_SIZE - 1) {
+            last_added_position = -1;
```

Comment:
maybe set to zero and increment after adding ? Then you could use unsigned types as indexes too

Reply:
discussed offline